File: hw-5B/C++/A-prefix-sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>

using namespace std;

typedef long long ll;

vector <ll> buildPrefixSum(vector <ll> nums){
    vector <ll> prefixsum(nums.size()+1, 0);
    for (int i=1; i<nums.size()+1; i++){
        prefixsum[i] = prefixsum[i-1] + nums[i-1];
    }
    return prefixsum;
}


int main(){

    // Calculate prefix sums for segments [l, r] of the array

    long n, k;
    long l, r;
    ll x;
    vector <ll> arr, prefixsum;
    cin >> n >> k;
    for (long i=0; i<n; i++){
        cin >> x;
        arr.push_back(x);
    }
    prefixsum = buildPrefixSum(arr);
    for (long i=0; i<k; i++){
        cin >> l >> r;
        cout << prefixsum[r] - prefixsum[l-1] << endl;
    }
    
    return 0;
}

File: hw-5B/Python/A-prefix-sum.py
def build_prefixsum(nums):
    prefixsum = [0] * (len(nums) + 1)
    for i in range(1, len(nums) + 1):
        prefixsum[i] = prefixsum[i-1] + nums[i-1]
    return prefixsum


def segment_sum(prefixsum, l, r):
    return prefixsum[r] - prefixsum[l-1]


if __name__ == '__main__':

    """ Calculate prefix sums for segments [l, r] of the array"""

    n, q = map(int, input().split())
    arr = [int(x) for x in input().split()]
    segments = []
    for i in range(q):
        l, r = map(int, input().split())
        segments.append((l, r))
    prefixsum = build_prefixsum(arr)
    for l, r in segments:
        print(segment_sum(prefixsum, l, r))

